content/2016-08-17-patterns.rst

```diff
@@ -0,0 +1,124 @@
+       fmt.Printf("%s\t%d\n", name, ages[name])
+   }
+
+Peter Norvig calls this an `informal design pattern
+<http://norvig.com/design-patterns/>`_: something referred to by name
+("iterate through items in a map in order of keys") and re-implemented from
+scratch each time it's needed.
+
+Informal patterns have their place, but we can often do better. Patterns are
+one way of expressing "how to" knowledge [2]_ but we have another way: code.
+Source code is a *formal* expression of "how to" knowledge that we can
+execute, test, manipulate, verify, compose, and re-use.
+
+Encoding "how to" knowledge is largely what programming *is* [3]_. We talk
+about replacing people with programs precisely because we take the knowledge
+about how to do their job and encode it such that even a machine can
+understand it.
+
+So how can we encode the knowledge of iterating through the items in a map in
+order of keys?
+
+We can start by copying Peter Norvig and reach for a dynamic programming
+language, such as Python:
+
+.. code-block:: python
+
+   names = []
+   for name in ages:
+       names.append(name)
+   names.sort()
+   for name in names:
+       print("{}\t{}".format(name, ages[name]))
+
+This is a very literal translation of the first snippet. A more idiomatic
+approach would look like:
+
+.. code-block:: python
+
+   names = sorted(ages.keys())
+   for name in names:
+       print("{}\t{}".format(name, ages[name])
+
+To turn this into a formal pattern, we need to extract a function that takes a
+map and returns a list of pairs of ``(key, value)`` in sorted order, like so:
+
+.. code-block:: python
+
+   def sorted_items(d):
+       result = []
+       sorted_keys = sorted(d.keys())
+       for k in sorted_keys:
+           result.append((k, d[k]))
+       return result
+
+   for name, age in sorted_items(ages):
+       print("{}\t{}".format(name, ages[name]))
+
+The pattern has become a function. Instead of a name or a description, it has
+an *identifier*, a True Name that gives us power over the thing. When we
+invoke it we don't need to comment our code to indicate that we are using a
+pattern because the name ``sorted_items`` makes it clear. If we choose, we can
+test it, optimize it, or perhaps even prove its correctness.
+
+If we figure out a better way of doing it, such as:
+
+.. code-block:: python
+
+   def sorted_items(d):
+       return [(k, d[k]) for k in sorted(d.keys())]
+
+Then we only have to change one place.
+
+It was being able to write code like this that drew me towards Python and away
+from Java, way back in 2001. It wasn't just that I could get more done in
+fewer lines—although that helped—it was that I could write what I meant.
+
+Of course, these days I'd much rather write:
+
+.. code-block:: haskell
+
+   import qualified Data.HashMap as Map
+   import Data.Maybe (fromJust)
+
+   sortedItems :: (Hashable k, Ord k) => Map.Map k v -> [(k, v)]
+   sortedItems d = [(k, fromJust k d) | k <- sort (Map.keys d)]
```

Comment:
Quite! Changed.

FWIW, using `Data.Map` would give me `toAscList`, which I think takes advantage of the internal sorting of the tree map, but this is more closely parallel to the Go & Python versions.